File: implementations/python/core/data_models.py
import sys
import uuid
from abc import ABC, abstractmethod
from dataclasses import dataclass
from decimal import Decimal
from enum import Enum
from typing import Any, Callable, Optional, Tuple

import pandas as pd
from pydantic import BaseModel, Field, model_validator


class ItemFrequency(str, Enum):
    """Enum for income/expense item frequency."""

    RECURRING = "recurring"
    ONE_TIME = "one-time"


class TimeUnit(str, Enum):
    """Time unit for amount input."""

    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    ANNUALLY = "annually"


class LiquidityLevel(str, Enum):
    """Asset liquidity level for cash flow strategy optimization."""

    HIGH = "high"  # Cash, checking accounts, money market funds - immediately available
    MEDIUM = (
        "medium"  # Stocks, ETFs - physically liquid but psychologically less frequent
    )
    LOW = "low"  # CDs, bonds - early withdrawal may incur penalties


class AssetClass(BaseModel):
    """Individual asset class in portfolio."""

    name: str = Field(
        ..., description="Asset class name (e.g., 'stocks', 'bonds', 'savings', 'cash')"
    )
    display_name: str = Field(
        default="", description="Original display name for UI (auto-filled if empty)"
    )

    @model_validator(mode="before")
    @classmethod
    def normalize_name(cls, data: Any) -> Any:
        """Normalize asset name for internal consistency"""
        if isinstance(data, dict) and "name" in data:
            original_name = str(data["name"]).strip()
            # Store original for display if display_name not provided
            if not data.get("display_name"):
                data["display_name"] = original_name
            # Normalize: lowercase + collapse multiple spaces to single space
            import re  # noqa: F401 - local import for validation

            normalized_name = re.sub(r"\s+", " ", original_name.lower().strip())
            data["name"] = normalized_name
        return data

    allocation_percentage: float = Field(
        ..., description="Allocation percentage", ge=0.0, le=100.0
    )
    expected_return: float = Field(
        ..., description="Expected annual return rate (%) - after tax"
    )
    volatility: float = Field(
        0.0,
        description=(
            "Expected volatility/risk (%) - ONLY used for Monte Carlo simulations"
        ),
    )
    liquidity_level: LiquidityLevel = Field(
        LiquidityLevel.MEDIUM,
        description="Asset liquidity level for cash flow optimization",
    )


class PortfolioConfiguration(BaseModel):
    """Investment portfolio configuration."""

    asset_classes: list[AssetClass] = Field(
        default_factory=lambda: [
            AssetClass(
                name="Stocks",  # Will be normalized to "stocks"
                display_name="Stocks",
                allocation_percentage=30.0,
                expected_return=5.0,
                volatility=15.0,
                liquidity_level=LiquidityLevel.MEDIUM,
            ),
            AssetClass(
                name="Bonds",  # Will be normalized to "bonds"
                display_name="Bonds",
                allocation_percentage=0.0,
                expected_return=3.0,
                volatility=5.0,
                liquidity_level=LiquidityLevel.LOW,
            ),
            AssetClass(
                name="Savings",  # Will be normalized to "savings"
                display_name="Savings",
                allocation_percentage=60.0,
                expected_return=1.0,
                volatility=5.0,
                liquidity_level=LiquidityLevel.LOW,
            ),
            AssetClass(
                name="Cash",  # Will be normalized to "cash"
                display_name="Cash",
                allocation_percentage=10.0,
                expected_return=0.0,
                volatility=1.0,
                liquidity_level=LiquidityLevel.HIGH,
            ),
        ],
        description="List of asset classes in the portfolio",
    )

    enable_rebalancing: bool = Field(
        True, description="Whether to rebalance portfolio annually"
    )

    @model_validator(mode="after")
    def validate_allocation_sum(self) -> "PortfolioConfiguration":
        """Validate that asset allocations sum to 100% (strict validation)"""
        total = sum(asset.allocation_percentage for asset in self.asset_classes)
        # Use sys.float_info.epsilon for machine precision tolerance
        tolerance = sys.float_info.epsilon

        if abs(total - 100.0) > tolerance:
            raise ValueError(
                f"Asset allocation percentages must sum to exactly 100%, "
                f"got {total}% (difference: {total - 100.0}%)"
            )
        return self

    @model_validator(mode="after")
    def validate_unique_asset_names(self) -> "PortfolioConfiguration":
        """Validate that asset names are unique within portfolio (case-insensitive)"""
        names = [
            asset.name for asset in self.asset_classes
        ]  # Already normalized to lowercase
        if len(names) != len(set(names)):
            duplicates = [name for name in names if names.count(name) > 1]
            unique_duplicates = list(set(duplicates))
            # Show display names in error for better UX
            display_names = [
                asset.display_name
                for asset in self.asset_classes
                if asset.name in unique_duplicates
            ]
            raise ValueError(
                f"Asset names must be unique within portfolio (case-insensitive). "
                f"Duplicate names found: {display_names}"
            )
        return self


class UserProfile(BaseModel):
    """Data model for the user's profile."""

    birth_year: int = Field(..., description="User's birth year")
    expected_fire_age: int = Field(50, description="User's expected FIRE age")
    legal_retirement_age: int = Field(
        65, description="Legal retirement age (when eligible for government pension)"
    )
    life_expectancy: int = Field(85, description="User's life expectancy")
    current_net_worth: float = Field(
        0.0, description="User's current net worth (after-tax value)"
    )
    inflation_rate: float = Field(3.0, description="Expected annual inflation rate (%)")

    # Safety buffer configuration
    safety_buffer_months: float = Field(
        12.0,
        description="Safety buffer in months of annual expenses (default: 12 months)",
    )

    # Investment configuration
    portfolio: PortfolioConfiguration = Field(
        default_factory=lambda: PortfolioConfiguration(enable_rebalancing=True)
    )

    @property
    def current_age(self) -> int:
        """Calculate current age from birth year"""
        from datetime import datetime

        return datetime.now().year - self.birth_year

    # Birth year validation
    @model_validator(mode="after")
    def validate_birth_year(self) -> "UserProfile":
        """Validate birth year is reasonable"""
        from datetime import datetime

        current_year = datetime.now().year
        if self.birth_year < 1950 or self.birth_year > current_year:
            raise ValueError(
                f"Birth year must be between 1950 and {current_year}, "
                f"got {self.birth_year}"
            )
        return self

    # Age validation using chain comparison
    @model_validator(mode="after")
    def validate_age_progression(self) -> "UserProfile":
        """
        Validate ages follow logic: current <= fire <= retirement <= life_expectancy
        """
        current = self.current_age
        fire = self.expected_fire_age
        retirement = self.legal_retirement_age
        life = self.life_expectancy

        if not (current <= fire <= retirement <= life):
            raise ValueError(
                f"Ages must follow progression: current_age({current}) <= "
                f"expected_fire_age({fire}) <= legal_retirement_age({retirement}) <= "
                f"life_expectancy({life})"
            )
        return self


class IncomeExpenseItem(BaseModel):
    """Data model for a single income or expense item."""

    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    name: str

    # Amount and time unit (after-tax values)
    after_tax_amount_per_period: float = Field(
        ..., description="After-tax amount per time period"
    )
    time_unit: TimeUnit = Field(
        TimeUnit.ANNUALLY, description="Time unit for the amount"
    )

    # Frequency control
    frequency: ItemFrequency = Field(
        ItemFrequency.RECURRING, description="Frequency type"
    )
    interval_periods: int = Field(
        1,
        description=(
            "Interval in time_unit periods "
            "(e.g., 6 for every 6 months if time_unit is monthly)"
        ),
        gt=0,
    )

    # Time range
    start_age: int = Field(..., description="The age this item starts")
    end_age: Optional[int] = Field(
        None, description="The age this item ends (for recurring items)"
    )

    # Growth rate (always annualized, after-tax)
    annual_growth_rate: float = Field(
        0.0, description="Annual growth rate (%) - after tax"
    )

    # Additional fields for categorization and validation
    is_income: bool = Field(..., description="True for income, False for expense")
    category: Optional[str] = Field(
        None, description="Category for grouping (e.g., 'Housing', 'Transportation')"
    )


# =============================================================================
# Calculation Results
# =============================================================================


@dataclass
class YearlyState:
    """Complete state for a single year in FIRE calculation."""

    age: int
    year: int

    # Cash flows (from input table)
    total_income: float
    total_expense: float
    net_cash_flow: float  # income - expense

    # Portfolio state
    portfolio_value: Decimal
    investment_return: Decimal  # Annual return from portfolio

    # Sustainability metrics (core logic)
    safety_buffer_amount: float  # Required safety buffer based on annual expenses
    is_sustainable: (
        bool  # True if net worth can remain above safety buffer through life expectancy
    )

    # Traditional FIRE metrics (optional reference)
    fire_number: float  # 25x annual expenses (4% rule) - for reference only
    fire_progress: float  # portfolio_value / fire_number (0-1+) - for reference only


class FIRECalculationResult(BaseModel):
    """Complete FIRE calculation result."""

    # Core sustainability results
    is_fire_achievable: bool = Field(
        ..., description="Whether FIRE is achievable based on net worth sustainability"
    )
    fire_net_worth: float = Field(..., description="Net worth at expected FIRE age")

    # Net worth trajectory analysis
    min_net_worth_after_fire: float = Field(
        ..., description="Minimum net worth during retirement phase"
    )
    final_net_worth: float = Field(
        ..., description="Net worth at end of life expectancy"
    )

    # Safety buffer analysis
    safety_buffer_months: float = Field(
        ..., description="Configured safety buffer in months"
    )
    min_safety_buffer_ratio: float = Field(
        ..., description="Minimum ratio of net worth to safety buffer (worst case)"
    )

    # Detailed yearly results
    yearly_results: list[YearlyState] = Field(
        ..., description="Year-by-year calculation results"
    )

    # Traditional FIRE metrics (for reference)
    traditional_fire_number: float = Field(
        ..., description="Traditional 4% rule FIRE number (25x expenses)"
    )
    traditional_fire_achieved: bool = Field(
        ..., description="Whether traditional 4% FIRE is achieved"
    )

    # Risk analysis (will be populated by Monte Carlo)
    fire_success_probability: Optional[float] = Field(
        None, description="Monte Carlo success probability (0-1)"
    )

    # Summary statistics
    total_years_simulated: int = Field(..., description="Total years in simulation")
    retirement_years: int = Field(
        ..., description="Years from FIRE age to life expectancy"
    )


# =============================================================================
# Simulation and Events
# =============================================================================


@dataclass
class BlackSwanEvent(ABC):
    """Base class for black swan events with self-contained calculation logic."""

    event_id: str  # Unique identifier (e.g., 'financial_crisis')
    annual_probability: float  # Annual occurrence probability (0-1)
    duration_years: int  # Duration of impact in years
    recovery_factor: float  # Recovery factor (0-1, 1 means full recovery)
    age_range: Tuple[int, int] = (18, 100)  # Applicable age range

    @abstractmethod
    def apply_impact(
        self, df: pd.DataFrame, year_idx: int, recovery_multiplier: float = 1.0
    ) -> pd.DataFrame:
        """Apply this event's impact to a specific year.

        Args:
            df: DataFrame to modify
            year_idx: Year index in the DataFrame
            recovery_multiplier: Multiplier for recovery (< 1.0 for ongoing events)

        Returns:
            Modified DataFrame
        """

    def get_display_name(self, i18n_func: Optional[Callable[[str], str]] = None) -> str:
        """Get localized display name for this event.

        Args:
            i18n_func: i18n function, if None returns event_id

        Returns:
            Localized event name or event_id as fallback
        """
        if i18n_func:
            return i18n_func(f"events.{self.event_id}.name")
        return self.event_id


class SimulationSettings(BaseModel):
    """Monte Carlo simulation settings."""

    num_simulations: int = Field(1000, description="Number of simulations to run")
    confidence_level: float = Field(
        0.95, description="Confidence level for results", ge=0.5, le=0.99
    )
    include_black_swan_events: bool = Field(
        True,
        description="Whether to include black swan events (including market crashes)",
    )

    # Base variation parameters (daily fluctuations, excluding major events)
    # Both use normal distribution with mean=1.0 and configurable standard deviation
    income_base_volatility: float = Field(
        0.1, description="Base income volatility (standard deviation)", ge=0.0, le=1.0
    )
    income_minimum_factor: float = Field(
        0.1, description="Minimum income factor (safety net threshold)", ge=0.01, le=1.0
    )

    expense_base_volatility: float = Field(
        0.05, description="Base expense volatility (standard deviation)", ge=0.0, le=1.0
    )
    expense_minimum_factor: float = Field(
        0.5,
        description="Minimum expense factor (prevents expenses going too low)",
        ge=0.1,
        le=1.0,
    )

File: implementations/typescript/src/core/data_models.ts
/**
 * FIRE Balance Calculator Data Models - Direct port from Python data_models.py
 *
 * This file contains all the data models with exact 1:1 mapping to Python Pydantic models.
 * All validation logic, defaults, and behaviors are preserved.
 */

import { v4 as uuidv4 } from "uuid";
import Decimal from "decimal.js";

// =============================================================================
// Enums - Direct mapping from Python
// =============================================================================

/**
 * Enum for income/expense item frequency
 * Direct mapping from Python ItemFrequency
 */
export type ItemFrequency = "recurring" | "one-time";

/**
 * Time unit for amount input
 * Direct mapping from Python TimeUnit
 */
export type TimeUnit = "monthly" | "quarterly" | "annually";

/**
 * Asset liquidity level for cash flow strategy optimization
 * Direct mapping from Python LiquidityLevel
 */
export type LiquidityLevel = "high" | "medium" | "low";

// =============================================================================
// Asset and Portfolio Classes - Direct mapping from Python
// =============================================================================

/**
 * Individual asset class in portfolio
 * Direct TypeScript equivalent of Python's AssetClass model
 */
export interface AssetClass {
  /** Asset class name (normalized to lowercase, e.g., 'stocks', 'bonds') */
  name: string;

  /** Original display name for UI (auto-filled if empty) */
  display_name: string;

  /** Allocation percentage (0-100) */
  allocation_percentage: Decimal;

  /** Expected annual return rate (%) - after tax */
  expected_return: Decimal;

  /** Expected volatility/risk (%) - ONLY used for Monte Carlo simulations */
  volatility: Decimal;

  /** Asset liquidity level for cash flow optimization */
  liquidity_level: LiquidityLevel;
}

/**
 * Normalize asset name for internal consistency
 * Direct port of Python's normalize_name validator
 */
export function normalizeAssetClass(data: Partial<AssetClass>): AssetClass {
  if (data.name) {
    const originalName = String(data.name).trim();
    // Store original for display if display_name not provided
    if (!data.display_name) {
      data.display_name = originalName;
    }
    // Normalize: lowercase + collapse multiple spaces to single space
    const normalizedName = originalName
      .toLowerCase()
      .trim()
      .replace(/\s+/g, " ");
    data.name = normalizedName;
  }

  return {
    name: data.name || "",
    display_name: data.display_name || "",
    allocation_percentage: new Decimal(data.allocation_percentage || 0),
    expected_return: new Decimal(data.expected_return || 0),
    volatility: new Decimal(data.volatility || 0),
    liquidity_level: data.liquidity_level || "medium",
  };
}

/**
 * Investment portfolio configuration
 * Direct TypeScript equivalent of Python's PortfolioConfiguration model
 */
export interface PortfolioConfiguration {
  /** List of asset classes in the portfolio */
  asset_classes: AssetClass[];

  /** Whether to rebalance portfolio annually */
  enable_rebalancing: boolean;
}

/**
 * Default portfolio configuration matching Python exactly
 * Direct mapping from Python's default_factory
 */
export const DEFAULT_PORTFOLIO: PortfolioConfiguration = {
  asset_classes: [
    {
      name: "stocks", // Will be normalized to "stocks"
      display_name: "Stocks",
      allocation_percentage: new Decimal(20.0),
      expected_return: new Decimal(5.0),
      volatility: new Decimal(15.0),
      liquidity_level: "medium",
    },
    {
      name: "bonds", // Will be normalized to "bonds"
      display_name: "Bonds",
      allocation_percentage: new Decimal(0.0),
      expected_return: new Decimal(3.0),
      volatility: new Decimal(5.0),
      liquidity_level: "low",
    },
    {
      name: "savings", // Will be normalized to "savings"
      display_name: "Savings",
      allocation_percentage: new Decimal(60.0),
      expected_return: new Decimal(1.0),
      volatility: new Decimal(5.0),
      liquidity_level: "low",
    },
    {
      name: "cash", // Will be normalized to "cash"
      display_name: "Cash",
      allocation_percentage: new Decimal(20.0),
      expected_return: new Decimal(0.0),
      volatility: new Decimal(1.0),
      liquidity_level: "high",
    },
  ],
  enable_rebalancing: true,
};

/**
 * Validate that asset allocations sum to 100% (strict validation)
 * Direct port of Python's validate_allocation_sum
 */
export function validateAllocationSum(portfolio: PortfolioConfiguration): void {
  const total = portfolio.asset_classes.reduce(
    (sum, asset) => sum.add(asset.allocation_percentage),
    new Decimal(0),
  );
  // Use Decimal precision tolerance
  const tolerance = new Decimal("1e-10");

  if (total.sub(100).abs().gt(tolerance)) {
    throw new Error(
      `Asset allocation percentages must sum to exactly 100%, ` +
        `got ${total.toString()}% (difference: ${total.sub(100).toString()}%)`,
    );
  }
}

/**
 * Validate that asset names are unique within portfolio (case-insensitive)
 * Direct port of Python's validate_unique_asset_names
 */
export function validateUniqueAssetNames(
  portfolio: PortfolioConfiguration,
): void {
  const names = portfolio.asset_classes.map((asset) => asset.name); // Already normalized to lowercase
  if (names.length !== new Set(names).size) {
    const duplicates = names.filter(
      (name, index) => names.indexOf(name) !== index,
    );
    const uniqueDuplicates = [...new Set(duplicates)];
    // Show display names in error for better UX
    const displayNames = portfolio.asset_classes
      .filter((asset) => uniqueDuplicates.includes(asset.name))
      .map((asset) => asset.display_name);
    throw new Error(
      `Asset names must be unique within portfolio (case-insensitive). ` +
        `Duplicate names found: ${displayNames.join(", ")}`,
    );
  }
}

/**
 * Create and validate a portfolio configuration
 * Combines normalization and validation like Python Pydantic
 */
export function createPortfolioConfiguration(data: {
  asset_classes?: (Partial<AssetClass> & {
    allocation_percentage?: number | Decimal;
    expected_return?: number | Decimal;
    volatility?: number | Decimal;
  })[];
  enable_rebalancing?: boolean;
}): PortfolioConfiguration {
  // Normalize all asset classes
  const normalizedAssets = (
    data.asset_classes || DEFAULT_PORTFOLIO.asset_classes
  ).map(normalizeAssetClass);

  const portfolio: PortfolioConfiguration = {
    asset_classes: normalizedAssets,
    enable_rebalancing: data.enable_rebalancing ?? true,
  };

  // Run validations
  validateAllocationSum(portfolio);
  validateUniqueAssetNames(portfolio);

  return portfolio;
}

// =============================================================================
// User Profile - Direct mapping from Python
// =============================================================================

/**
 * Data model for the user's profile
 * Direct TypeScript equivalent of Python's UserProfile model
 */
export interface UserProfile {
  /** User's birth year */
  birth_year: number;

  /** User's expected FIRE age */
  expected_fire_age: number;

  /** Legal retirement age (when eligible for government pension) */
  legal_retirement_age: number;

  /** User's life expectancy */
  life_expectancy: number;

  /** User's current net worth (after-tax value) */
  current_net_worth: Decimal;

  /** Expected annual inflation rate (%) */
  inflation_rate: Decimal;

  /** Safety buffer in months of annual expenses (default: 6 months) */
  safety_buffer_months: Decimal;

  /** Investment portfolio configuration */
  portfolio: PortfolioConfiguration;
}

/**
 * Calculate current age from birth year
 * Direct port of Python's current_age property
 */
export function getCurrentAge(birthYear: number): number {
  return new Date().getFullYear() - birthYear;
}

/**
 * Validate birth year is reasonable
 * Direct port of Python's validate_birth_year
 */
export function validateBirthYear(birthYear: number): void {
  const currentYear = new Date().getFullYear();
  if (birthYear < 1950 || birthYear > currentYear) {
    throw new Error(
      `Birth year must be between 1950 and ${currentYear}, got ${birthYear}`,
    );
  }
}

/**
 * Validate ages follow logic: current <= fire <= retirement <= life_expectancy
 * Direct port of Python's validate_age_progression
 */
export function validateAgeProgression(profile: UserProfile): void {
  const current = getCurrentAge(profile.birth_year);
  const fire = profile.expected_fire_age;
  const retirement = profile.legal_retirement_age;
  const life = profile.life_expectancy;

  if (!(current <= fire && fire <= retirement && retirement <= life)) {
    throw new Error(
      `Ages must follow progression: current_age(${current}) <= ` +
        `expected_fire_age(${fire}) <= legal_retirement_age(${retirement}) <= ` +
        `life_expectancy(${life})`,
    );
  }
}

/**
 * Create and validate a user profile
 * Combines validation like Python Pydantic
 */
export function createUserProfile(
  data: Partial<UserProfile> & {
    current_net_worth?: number | Decimal;
    inflation_rate?: number | Decimal;
    safety_buffer_months?: number | Decimal;
  },
): UserProfile {
  const profile: UserProfile = {
    birth_year: data.birth_year || 1990,
    expected_fire_age: data.expected_fire_age || 50,
    legal_retirement_age: data.legal_retirement_age || 65,
    life_expectancy: data.life_expectancy || 85,
    current_net_worth: new Decimal(data.current_net_worth || 0.0),
    inflation_rate: new Decimal(data.inflation_rate || 3.0),
    safety_buffer_months: new Decimal(data.safety_buffer_months || 6),
    portfolio: data.portfolio || createPortfolioConfiguration({}),
  };

  // Run validations
  validateBirthYear(profile.birth_year);
  validateAgeProgression(profile);

  return profile;
}

// =============================================================================
// Income/Expense Items - Direct mapping from Python
// =============================================================================

/**
 * Data model for a single income or expense item
 * Direct TypeScript equivalent of Python's IncomeExpenseItem model
 */
export interface IncomeExpenseItem {
  /** Unique identifier */
  id: string;

  /** Item name */
  name: string;

  /** After-tax amount per time period */
  after_tax_amount_per_period: Decimal;

  /** Time unit for the amount */
  time_unit: TimeUnit;

  /** Frequency type */
  frequency: ItemFrequency;

  /** Interval in time_unit periods */
  interval_periods: number;

  /** The age this item starts */
  start_age: number;

  /** The age this item ends (for recurring items) */
  end_age?: number;

  /** Annual growth rate (%) - after tax */
  annual_growth_rate: Decimal;

  /** True for income, False for expense */
  is_income: boolean;

  /** Category for grouping */
  category?: string;
}

/**
 * Create an income/expense item with validation
 * Mimics Python's default behavior and validation
 */
export function createIncomeExpenseItem(
  data: Partial<IncomeExpenseItem> & {
    after_tax_amount_per_period?: number | Decimal;
    annual_growth_rate?: number | Decimal;
  },
): IncomeExpenseItem {
  const item: IncomeExpenseItem = {
    id: data.id || uuidv4(),
    name: data.name || "",
    after_tax_amount_per_period: new Decimal(
      data.after_tax_amount_per_period || 0,
    ),
    time_unit: data.time_unit || "annually",
    frequency: data.frequency || "recurring",
    interval_periods: data.interval_periods ?? 1,
    start_age: data.start_age || 0,
    end_age: data.end_age,
    annual_growth_rate: new Decimal(data.annual_growth_rate || 0.0),
    is_income: data.is_income || false,
    category: data.category,
  } as IncomeExpenseItem;

  // Validate interval_periods > 0 (matches Python's gt=0 constraint)
  if (item.interval_periods <= 0) {
    throw new Error("interval_periods must be greater than 0");
  }

  return item;
}

// =============================================================================
// Calculation Results - Direct mapping from Python
// =============================================================================

/**
 * Complete state for a single year in FIRE calculation
 * Direct TypeScript equivalent of Python's YearlyState dataclass
 */
export interface YearlyState {
  /** User's age in this year */
  age: number;

  /** Calendar year */
  year: number;

  // Cash flows (from input table)
  /** Total income for the year */
  total_income: Decimal;

  /** Total expenses for the year */
  total_expense: Decimal;

  /** Net cash flow (income - expense) */
  net_cash_flow: Decimal;

  // Financial state
  /** Investment portfolio value (always >= 0) */
  portfolio_value: Decimal;

  /** True net worth (can be negative, indicating debt) */
  net_worth: Decimal;

  /** Annual return from portfolio */
  investment_return: Decimal;

  // Sustainability metrics (core logic)
  /** True if net worth can remain above safety buffer through life expectancy */
  is_sustainable: boolean;

  // Traditional FIRE metrics (optional reference)
  /** Traditional 4% rule FIRE number (25x annual expenses) - for reference only */
  fire_number: Decimal;

  /** Portfolio value / fire_number (0-1+) - for reference only */
  fire_progress: Decimal;
}

/**
 * Complete FIRE calculation result
 * Direct TypeScript equivalent of Python's FIRECalculationResult model
 */
export interface FIRECalculationResult {
  // Core sustainability results
  /** Whether FIRE is achievable based on net worth sustainability */
  is_fire_achievable: boolean;

  /** Net worth at expected FIRE age */
  fire_net_worth: Decimal;

  // Net worth trajectory analysis
  /** Minimum net worth during retirement phase */
  min_net_worth_after_fire: Decimal;

  /** Net worth at end of life expectancy */
  final_net_worth: Decimal;

  // Safety buffer analysis
  /** Configured safety buffer in months */
  safety_buffer_months: Decimal;

  /** Minimum ratio of net worth to safety buffer (worst case) */
  min_safety_buffer_ratio: Decimal;

  // Detailed yearly results
  /** Year-by-year calculation results */
  yearly_results: YearlyState[];

  // Traditional FIRE metrics (for reference)
  /** Traditional 4% rule FIRE number (25x expenses) */
  traditional_fire_number: Decimal;

  /** Whether traditional 4% FIRE is achieved */
  traditional_fire_achieved: boolean;

  // Risk analysis (will be populated by Monte Carlo)
  /** Monte Carlo success probability (0-1) */
  fire_success_probability?: Decimal;

  // Summary statistics
  /** Total years in simulation */
  total_years_simulated: number;

  /** Years from FIRE age to life expectancy */
  retirement_years: number;
}

// =============================================================================
// Simulation and Events - Direct mapping from Python
// =============================================================================

/**
 * Base interface for black swan events
 * TypeScript equivalent of Python's BlackSwanEvent abstract class
 */
export interface BlackSwanEvent {
  /** Unique identifier (e.g., 'financial_crisis') */
  event_id: string;

  /** Annual occurrence probability (0-1) */
  annual_probability: Decimal;

  /** Duration of impact in years */
  duration_years: number;

  /** Recovery factor (0-1, 1 means full recovery) */
  recovery_factor: Decimal;

  /** Applicable age range */
  age_range: [number, number];
}

/**
 * Monte Carlo simulation settings
 * Direct TypeScript equivalent of Python's SimulationSettings model
 */
export interface SimulationSettings {
  /** Number of simulations to run */
  num_simulations: number;

  /** Confidence level for results (0.5-0.99) */
  confidence_level: Decimal;

  /** Whether to include black swan events (including market crashes) */
  include_black_swan_events: boolean;

  // Base variation parameters (daily fluctuations, excluding major events)
  /** Base income volatility (standard deviation) */
  income_base_volatility: Decimal;

  /** Minimum income factor (safety net threshold) */
  income_minimum_factor: Decimal;

  /** Base expense volatility (standard deviation) */
  expense_base_volatility: Decimal;

  /** Minimum expense factor (prevents expenses going too low) */
  expense_minimum_factor: Decimal;
}

/**
 * Default simulation settings matching Python defaults
 */
export const DEFAULT_SIMULATION_SETTINGS: SimulationSettings = {
  num_simulations: 1000,
  confidence_level: new Decimal(0.95),
  include_black_swan_events: true,
  income_base_volatility: new Decimal(0.1),
  income_minimum_factor: new Decimal(0.1),
  expense_base_volatility: new Decimal(0.05),
  expense_minimum_factor: new Decimal(0.5),
};

/**
 * Create and validate simulation settings
 * Mimics Python Pydantic validation behavior
 */
export function createSimulationSettings(
  data: Partial<SimulationSettings> & {
    confidence_level?: number | Decimal;
    income_base_volatility?: number | Decimal;
    income_minimum_factor?: number | Decimal;
    expense_base_volatility?: number | Decimal;
    expense_minimum_factor?: number | Decimal;
  },
): SimulationSettings {
  const settings: SimulationSettings = {
    num_simulations:
      data.num_simulations ?? DEFAULT_SIMULATION_SETTINGS.num_simulations,
    confidence_level: new Decimal(
      data.confidence_level ?? DEFAULT_SIMULATION_SETTINGS.confidence_level,
    ),
    include_black_swan_events:
      data.include_black_swan_events ??
      DEFAULT_SIMULATION_SETTINGS.include_black_swan_events,
    income_base_volatility: new Decimal(
      data.income_base_volatility ??
        DEFAULT_SIMULATION_SETTINGS.income_base_volatility,
    ),
    income_minimum_factor: new Decimal(
      data.income_minimum_factor ??
        DEFAULT_SIMULATION_SETTINGS.income_minimum_factor,
    ),
    expense_base_volatility: new Decimal(
      data.expense_base_volatility ??
        DEFAULT_SIMULATION_SETTINGS.expense_base_volatility,
    ),
    expense_minimum_factor: new Decimal(
      data.expense_minimum_factor ??
        DEFAULT_SIMULATION_SETTINGS.expense_minimum_factor,
    ),
  };

  // Validate confidence_level range (matches Python ge=0.5, le=0.99)
  if (settings.confidence_level.lt(0.5) || settings.confidence_level.gt(0.99)) {
    throw new Error(
      `confidence_level must be between 0.5 and 0.99, got ${settings.confidence_level.toString()}`,
    );
  }

  // Validate volatility ranges (matches Python ge=0.0, le=1.0)
  if (
    settings.income_base_volatility.lt(0.0) ||
    settings.income_base_volatility.gt(1.0)
  ) {
    throw new Error(
      `income_base_volatility must be between 0.0 and 1.0, got ${settings.income_base_volatility.toString()}`,
    );
  }
  if (
    settings.expense_base_volatility.lt(0.0) ||
    settings.expense_base_volatility.gt(1.0)
  ) {
    throw new Error(
      `expense_base_volatility must be between 0.0 and 1.0, got ${settings.expense_base_volatility.toString()}`,
    );
  }

  // Validate minimum factors (matches Python ge=0.01, le=1.0 for income, ge=0.1, le=1.0 for expense)
  if (
    settings.income_minimum_factor.lt(0.01) ||
    settings.income_minimum_factor.gt(1.0)
  ) {
    throw new Error(
      `income_minimum_factor must be between 0.01 and 1.0, got ${settings.income_minimum_factor.toString()}`,
    );
  }
  if (
    settings.expense_minimum_factor.lt(0.1) ||
    settings.expense_minimum_factor.gt(1.0)
  ) {
    throw new Error(
      `expense_minimum_factor must be between 0.1 and 1.0, got ${settings.expense_minimum_factor.toString()}`,
    );
  }

  return settings;
}
